### PR TITLE
Use meshes for highlighted objects

### PR DESCRIPTION
**Pull request by RealBadAngel**
Unifies displaying selection boxes and halo, both are using meshes now, also it is moved out of mapblock mesh generation, so pointing something doesnt trigger updating block mesh.
Selection boxes and halo effect can be now easily textured with texture packs,
Halo works now with any kind of pointed object.

Ive removed also settings to change selection boxes colour and line thickness, those effects can be achieved by just adding selectionbox.png to texture pack with custom box or whatever the shape user needs

### CODE REVIEW

**Comment by red-001** [comment]
:+1:  Everything seems to work correctly. Tested on windows 7 with Visual Studio 2013.

**Comment by RealBadAngel** [comment]
@red-001 yeah, i will push this and some other tweaks asap

**Comment by paramat** [comment]
I guess this also fixes the random selection box that appears with your post-pricessing shaders?

I'm shocked that selectionbox updates the block mesh as it moves rapidly from node to node, the mesh updating must be a huge load.

**Comment by RealBadAngel** [comment]
@paramat, updating the mesh is true in case of node highlighting and cracks. selection boxes before were simple lines, not able to write to depth buffer so they were messing almost each pp effect that was using it

**Comment by paramat** [comment]
I hope with this PR selection boxes are still thin lines, anything else for me is overkill.

**Comment by RealBadAngel** [comment]
@paramat:
screenshot_20160206_001922
screenshot_20160206_002128
screenshot_20160206_002346
screenshot_20160206_002636
screenshot_20160206_002824
screenshot_20160206_002904

**Comment by paramat** [comment]
In my opinion: Halo should be very subtle and mostly colour-neutral to keep the item clearly visible, just a misty-white or how it is currently. I actualy think the current halo animation is unnecessary, would it improve performance if it was a static texture?
The selection box should be the minimum sized texture needed to create a thin line when with 1 pixel of black around the edge, i'm guessing 64px or 128px would be enough.

**Comment by kaadmy** [comment]
:+1:
Looks good ;)

**Comment by Fixer-007** [comment]
:+1:

**Comment by RealBadAngel** [comment]
Fixed all the issues i could find.

**Comment by paramat** [comment]
Looking at the selectionbox texture in Gimp i noticed the grey had some red colour in it, i reduced the colour saturation to zero and also converted it to indexed colour, file size dropped from 846b to 284b.
If an indexed version is usable here it is:
selectionboxgsi

**Comment by paramat** [comment]
The halo has changed from a nice featureless sky blue to turquoise with an ugly smoothed and hires dark border. I think the turquoise makes an unpleasant combination with nde colours.

**Comment by paramat** [comment]
At larger distances the lines of the selection box start to disappear, more of a problem in creative mode:

screenshot_20160206_063125

But also affects normal mode a little:

screenshot_20160206_063404

Maybe the solution would be thicker box lines.
I guess currently the box lines are constant thickness at any distance,

**Comment by kilbith** [comment]
This black & white selection box looks bad. Can't you be more clever by inverting the colors ? (i.e. selection borders turn darker on clearer nodes, clearer on darker nodes).

**Comment by RealBadAngel** [comment]
@kilbith those are textures on a mesh... i havent heard of such smart textures yet ;)

**Comment by BlockMen** [comment]
IMO its a bad idea to switch to a texture for the selection box at all. I'm fine with moving your "selection mesh" out of mesh generation, but I'm against removing the simple selection box we have currently.

**Comment by RealBadAngel** [comment]
@BlockMen simple selection boxes (ie not real 3d objects) cannot be used for deffered rendering and have to be removed, pick what you like better boxes or real lighting

**Comment by C1ffisme** [comment]
Looks really good!
:+1:

**Comment by RealBadAngel** [comment]
Ok, seems like i have found the solution that will allow to keep simple boxes and at the same time to write depth and normal data to respective buffers for deferred rendering, will update PR with it, but for it to be merged #3650 have to go in first.
So, overall changes of this PR gonna be:
- move halo mesh out of mapblock mesh gen (rather huge improvement)
- improved halo look (no more overlapping boxes when using more complex ones)
- clean the mess with passing selection boxes all around between game.cpp, hud.cpp and drawscene.cpp
- sticker Deferred Rendering Ready (TM) ;)

**Comment by paramat** [comment]
> but I'm against removing the simple selection box we have currently.

If possible i'd like to keep it as it is too, as long as it still does not update meshes.
The way the current box lines stay constant thickness at all distances is good, keeps them clear at large distance and subtle at small distance.
The reach of the player in creative mode is very useful for building.

> will update PR with it, but for it to be merged #3650 have to go in first.

3650 is this PR.

File: src/hud.cpp [hunk 1]
```diff
@@ -465,14 +486,54 @@ void Hud::drawCrosshair() {
...
+			m_selectionmesh->drop();
+			m_selectionmesh = NULL;
+		}
+		return;
+	} else {
+		if (m_selectionmesh) {
+			m_selectionmesh->drop();
+			m_selectionmesh = NULL;
+		}
+		static f32 txc[24] = {
```
Comment (nerzhul) [inline src/hud.cpp]: static const

File: src/hud.h [hunk 2]
```diff
@@ -124,12 +124,31 @@ class Hud {
...
 	void resizeHotbar();
 	void drawCrosshair();
-	void drawSelectionBoxes(std::vector<aabb3f> &hilightboxes);
+	void drawSelectionMesh();
+	void updateSelectionMesh(v3s16 camera_offset);
+	
+	std::vector<aabb3f> *getSelectionBoxes()
+	{ return &m_selectionboxes; }
+
+	void setHighlightedPos(v3f pos)
```
Comment (nerzhul) [inline src/hud.h]: const v3f&

File: src/hud.h [hunk 3]
```diff
@@ -124,12 +124,31 @@ class Hud {
...
+	{ m_highlighted_pos = pos; }
+	
+	v3f getHighlightedPos()
```
Comment (nerzhul) [inline src/hud.h]: v3f get..() const

File: src/hud.h [hunk 4]
```diff
@@ -124,12 +124,31 @@ class Hud {
...
+	{ return m_highlighted_pos; }
+
+	void setHighlightedMeshColor(video::SColor c)
```
Comment (nerzhul) [inline src/hud.h]: cosnt SColor &

File: src/hud.h [hunk 5]
```diff
@@ -124,12 +124,31 @@ class Hud {
...
+	{ m_highlightedmeshcolor = c; }
+
+	void setSelectionMeshScale(v3f scale)
```
Comment (nerzhul) [inline src/hud.h]: const v3f&

File: src/mesh.cpp [hunk 6]
```diff
@@ -407,6 +407,13 @@ scene::IMesh* cloneMesh(scene::IMesh *src_mesh)
 
 scene::IMesh* convertNodeboxNodeToMesh(ContentFeatures *f)
 {
+	std::vector<aabb3f> boxes = f->node_box.fixed;
+	return convertNodeboxesToMesh (boxes, NULL);
+}
+
+scene::IMesh* convertNodeboxesToMesh(std::vector<aabb3f> boxes, f32* txc)
```
Comment (nerzhul) [inline src/mesh.cpp]: const std::vector &

File: src/hud.cpp [hunk 7]
```diff
@@ -465,14 +486,54 @@ void Hud::drawCrosshair() {
...
+	u32 mc = m_selectionmesh->getMeshBufferCount();
+	for (u32 i = 0; i < mc; i++) {
+		scene::IMeshBuffer *buf = m_selectionmesh->getMeshBuffer(i);
+		video::SMaterial &material = buf->getMaterial();
+		driver->setMaterial(material);
+		driver->drawMeshBuffer(buf);
 	}
 }
 
+void Hud::updateSelectionMesh(v3s16 camera_offset)
```
Comment (nerzhul) [inline src/hud.cpp]: const v3s16 &

File: src/drawscene.h [hunk 8]
```diff
@@ -31,9 +31,9 @@ void draw_load_screen(const std::wstring &text, IrrlichtDevice *device,
 		bool clouds = true);
 
 void draw_scene(video::IVideoDriver *driver, scene::ISceneManager *smgr,
-		Camera &camera, Client &client, LocalPlayer *player, Hud &hud,
-		Mapper &mapper, gui::IGUIEnvironment *guienv,
-		std::vector<aabb3f> hilightboxes, const v2u32 &screensize,
-		video::SColor skycolor, bool show_hud, bool show_minimap);
+		Camera &camera, Client &client, LocalPlayer *player,
+		Hud &hud, Mapper &mapper, gui::IGUIEnvironment *guienv,
+		const v2u32 &screensize, video::SColor skycolor,
```
Comment (nerzhul) [inline src/drawscene.h]: const SColor &

File: src/hud.cpp (hunk 1 above)
Comment (RealBadAngel) [inline src/hud.cpp]: this cannot be neither static nor const

File: src/drawscene.h (hunk 8 above)
Comment (RealBadAngel) [inline src/drawscene.h]: wont even compile

File: src/hud.cpp (hunk 1 above)
Comment (RealBadAngel) [inline src/hud.cpp]: cant be const because of iterators later on, wont compile even

**Comment by RealBadAngel** [comment]
rebased and updated in #3662 , so closing

**Pull request closed without merging**